### PR TITLE
[WiWa] iP

### PR DESCRIPTION
**Pull request by snuckerzlol**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,146 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
 public class Duke {
+
     public static void main(String[] args) {
```
Comment (JoshHDs) [inline src/main/java/Duke.java]: Your Main method seems very long. Consider extracting out methods
https://se-education.org/guides/conventions/java/basic.html

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,146 @@
...
+        while (!curCommand.equals("bye")) {
+            curCommand = input.nextLine();
+            System.out.println(line);
+
+            String[] commands = curCommand.split(" ", 2);
+
+            // Command has multiple parameters
+            if (commands.length > 1) {
+                switch (commands[0]) {
+                    case "done":
```
Comment (JoshHDs) [inline src/main/java/Duke.java]: Perhaps switch and case must be aligned
https://se-education.org/guides/conventions/java/basic.html

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,146 @@
...
+                    default:
+                        System.out.println("Input error");
+                }
+            }
+            System.out.println(line);
+        }
+    }
+
+    // Checks if a string is a positive numeric value
+    public static int isPositiveNumeric(String str) {
```
Comment (JoshHDs) [inline src/main/java/Duke.java]: Perhaps change naming of method as it is not returning a boolean
https://se-education.org/guides/conventions/java/basic.html

**Review by JoshHDs — COMMENTED** [review]
Looks good!

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,146 @@
...
+    }
+
+    // Converts task to string
+    public static String taskToString(Task task) {
+        return "[" + task.getType() + "][" + (task.isDone() ? "X" : " ") + "] "
+                + task.getTask() + (task.getType() == 'T' ? "" : " (" + (task.getType() == 'D' ? "by: " : "at: ")+ task.getTrail() + ")");
+    }
+
+    // Prints newest task
+    public static void printNew(Task task, int size) {
```
Comment (JoshHDs) [inline src/main/java/Duke.java]: Perhaps you could give a clearer name for your method, such as printNewTask.
https://se-education.org/guides/conventions/java/basic.html

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,146 @@
...
+            return Integer.parseInt(str);
+        } catch(NumberFormatException e){
+            return -1;
+        }
+    }
+
+    // Converts task to string
+    public static String taskToString(Task task) {
+        return "[" + task.getType() + "][" + (task.isDone() ? "X" : " ") + "] "
+                + task.getTask() + (task.getType() == 'T' ? "" : " (" + (task.getType() == 'D' ? "by: " : "at: ")+ task.getTrail() + ")");
```
Comment (vincentlauhl) [inline src/main/java/Duke.java]: This parts seems a bit messy, perhaps you could add this as a method in your classes to make it easier to read

File: src/main/java/Duke.java [hunk 6]
```diff
@@ -1,10 +1,146 @@
...
+    public static void printNew(Task task, int size) {
+        System.out.println("Got it. I've added this task:\n " +
+                taskToString(task) +
+                "\nNow you have " + size + " tasks in the list.");
+    }
+
+    // Prints help message
+    public static void printHelp(String command) {
+        switch (command) {
+            case "done":
```
Comment (vincentlauhl) [inline src/main/java/Duke.java]: Perhaps you could align 'case' to the 'switch', as per the course standards

File: src/main/java/Duke.java [hunk 7]
```diff
@@ -1,10 +1,146 @@
...
     public static void main(String[] args) {
+        Scanner input = new Scanner(System.in);
         String logo = " ____        _        \n"
                 + "|  _ \\ _   _| | _____ \n"
                 + "| | | | | | | |/ / _ \\\n"
                 + "| |_| | |_| |   <  __/\n"
                 + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        String line = "________________________________________";
+        ArrayList<Task> taskList = new ArrayList<>();
```
Comment (vincentlauhl) [inline src/main/java/Duke.java]: Perhaps you should name this as tasks, as per the coding standards of the course

File: src/main/java/Duke.java [hunk 8]
```diff
@@ -1,10 +1,146 @@
...
+                    case "event":
+                        String[] evin = commands[1].split(" /at ", 2);
+                        if (evin.length > 1) {
+                            taskList.add(new Task(evin[0], evin[1], 'E', false));
+                            printNew(taskList.get(taskList.size() - 1), taskList.size());
+                        }
+                        else {
+                            System.out.println("Formatting error");
+                        }
+                        break;
```
Comment (vincentlauhl) [inline src/main/java/Duke.java]: perhaps you could add a layer of abstraction to this to make it look neater

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -1,10 +1,178 @@
...
+
+        while (!curCommand.equals("bye")) {
+            curCommand = input.nextLine();
+            System.out.println(line);
+            handleCommand(curCommand);
+            System.out.println(line);
+        }
     }
+
+    // Handle command
```
Comment (iamabhishek98) [inline src/main/java/Duke.java]: Such two-word comments arent particularly useful. Try to make your comments more descriptive

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,178 @@
...
+            handleSingleCommand(curCommand);
+        }
+    }
+
+    // Handle multiple commands
+    public static void handleMultiCommands(String[] commands) {
+        switch (commands[0]) {
+        case "done":
+            int doneIndex = getPositiveNumeric(commands[1]);
+            if (doneIndex > 0 && doneIndex <= tasks.size()) {
```
Comment (iamabhishek98) [inline src/main/java/Duke.java]: This check on the index value seems to have been repeated multiple times in this file. In line with the DRY (Don't Repeat Yourself) principle, you might want to abstract this logic out to a validation helper function.

File: src/main/java/Duke.java [hunk 11]
```diff
@@ -1,10 +1,178 @@
...
+            break;
+        case "todo":
+            tasks.add(new Task(commands[1], "", 'T', false));
+            MessageManager.printNewTask(tasks.get(tasks.size() - 1), tasks.size());
+            writeIntoFile();
+            break;
+        case "deadline":
+            String[] deadin = commands[1].split(" /by ", 2);
+            if (deadin.length > 1) {
+                tasks.add(new Task(deadin[0], deadin[1], 'D', false));
```
Comment (iamabhishek98) [inline src/main/java/Duke.java]: Avoid doing class initialization in the argument of another function. Store the new class in a variable and then pass that variable as the argument to the function

File: src/main/java/Duke.java [hunk 12]
```diff
@@ -1,10 +1,178 @@
...
+    public static void handleMultiCommands(String[] commands) {
+        switch (commands[0]) {
+        case "done":
+            int doneIndex = getPositiveNumeric(commands[1]);
+            if (doneIndex > 0 && doneIndex <= tasks.size()) {
+                tasks.get(doneIndex - 1).setDone(true);
+            }
+            System.out.println(doneIndex > 0 && doneIndex <= tasks.size()
+                    ? "Nice! I've marked this task as done:\n " + MessageManager.taskToString(tasks.get(doneIndex - 1))
+                    : "Formatting error");
```
Comment (iamabhishek98) [inline src/main/java/Duke.java]: Avoid using ternary operators. They can make your code hard to understand. My suggestion would be to change this logic to use if ... else if statements and then conditionally print. 
Also, try not to have computational logic in your print statements as the sole purpose of print statements is to print. Extract the operations inside the print statement to different lines and once you have the final result, print that instead.

File: src/main/java/Duke.java [hunk 13]
```diff
@@ -1,10 +1,178 @@
...
+            newFile.createNewFile();
+
+            FileWriter myWriter = new FileWriter(dir + "\\data\\duke.txt");
+
+            for (Task t : tasks) {
+                myWriter.write(t.getTask()
+                        + " / " + t.getTrail()
+                        + " / " + t.getType()
+                        + " / " + (t.isDone() ? 1 : 0)
+                        + "\n"
```
Comment (iamabhishek98) [inline src/main/java/Duke.java]: As mentioned earlier, extract such nested logic into separate lines.

File: src/main/java/Duke.java [hunk 14]
```diff
@@ -1,10 +1,178 @@
...
+            file.createNewFile();
+
+            FileReader fileReader = null;
+            fileReader = new FileReader(file);
+
+            BufferedReader bufferedreader = new BufferedReader(fileReader);
+            String strLine;
+
+            while ((strLine = bufferedreader.readLine()) != null) {
+                String[] saved = strLine.split(" / ", 4);
```
Comment (iamabhishek98) [inline src/main/java/Duke.java]: Avoid using magic numbers like 4 - extract them out to constants and name them appropriately using proper naming conventions

File: src/main/java/MessageManager.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    public static void printNewTask(Task task, int size) {
+        System.out.println("Got it. I've added this task:\n " +
+                taskToString(task) +
+                "\nNow you have " + size + " tasks in the list.");
+    }
+
+    // Converts task to string
+    public static String taskToString(Task task) {
+        return "[" + task.getType() + "][" + (task.isDone() ? "X" : " ") + "] "
+                + task.getTask() + (task.getType() == 'T' ? "" : " (" + (task.getType() == 'D' ? "by: " : "at: ")+ task.getTrail() + ")");
```
Comment (iamabhishek98) [inline src/main/java/MessageManager.java]: As mentioned before, avoid ternary operators. Also, consider extracting out magic strings like 'T' and 'D' to constants.

**Review by iamabhishek98 — COMMENTED** [review]
I have added my suggestions. Please note that although I have not repeated myself, what I have commented in one place applies to the rest of the codebase as well.
Also, consider adding more in-line comments to make your code more readable.

**Pull request closed without merging**